solutions/src/main/java/ru/sberbank/school/task10/ScalableThreadPool.java

```diff
@@ -0,0 +1,151 @@
+
+import lombok.NonNull;
+import ru.sberbank.school.task10.util.IdsManager;
+
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+
+public class ScalableThreadPool implements ThreadPool {
+    private final Queue<Runnable> taskQueue;
+
+    private final int min;
+    private final int max;
+    private volatile int freeThreads = 0;
+    private final List<ThreadWorker> threads;
+    private final List<ThreadWorker> garbageThreads;
+    private final IdsManager idsManager;
+
+    public ScalableThreadPool(int min, int max) {
+        if (min <= 0 || max <= 0 || min > max) {
+            throw new IllegalArgumentException();
+        }
+        this.taskQueue = new LinkedList<>();
+        this.min = min;
+        this.max = max;
+        this.idsManager = new IdsManager(max + 1);
+        this.threads = new ArrayList<>(min);
+        this.garbageThreads = new ArrayList<>();
+    }
+
+    @Override
+    public void start() {
+        if (!threads.isEmpty()) {
+            throw new IllegalStateException("Пул уже запущен!");
+        }
+        synchronized (threads) {
+            for (int i = 0; i < min; i++) {
+                ThreadWorker threadWorker = new ThreadWorker(null);
+                threads.add(threadWorker);
+                threadWorker.start();
+            }
+            freeThreads = min;
+        }
+    }
+
+    @Override
+    public synchronized void stopNow() {
+        if (threads.isEmpty()) {
+            throw new IllegalStateException("Пул еще не запущен!");
+        }
+        taskQueue.clear();
+        removeGarbageThreads();
+        for (ThreadWorker t : threads) {
+            t.interrupt();
+        }
+        threads.clear();
+        idsManager.reset();
+        freeThreads = min;
+    }
+
+    @Override
+    public synchronized void execute(@NonNull Runnable runnable) {
+        if (threads.isEmpty()) {
+            throw new IllegalStateException("Пул еще не запущен!");
+        }
+        synchronized (garbageThreads) {
+            if (!garbageThreads.isEmpty()) {
+                removeGarbageThreads();
+            }
+        }
+        if (freeThreads == 0 && threads.size() <= max) {
+            ThreadWorker threadWorker = new ThreadWorker(runnable);
+            threads.add(threadWorker);
+            threadWorker.start();
+            return;
+        }
+
+        synchronized (taskQueue) {
+            taskQueue.add(runnable);
+            taskQueue.notify();
+        }
+    }
+
+    private void removeGarbageThreads() {
+        Iterator<ThreadWorker> threadWorkerIterator = garbageThreads.iterator();
+        ThreadWorker threadWorker;
+        while (threadWorkerIterator.hasNext()) {
+            threadWorker = threadWorkerIterator.next();
+            idsManager.addFreeId(threadWorker.getThreadId());
+            threads.remove(threadWorker);
+        }
+        garbageThreads.clear();
+    }
+
+    @Override
```

Comment:
Overload methods should not be split. Previous overloaded method located at line &#39;63&#39;.

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a1371396728>*